### PR TITLE
Move `TyCtxt::mk_x` to `Ty::new_x` where applicable

### PR DESCRIPTION
**Pull request by BoxyUwU**
Part of rust-lang/compiler-team#616

turns out there's a lot of places we construct `Ty` this is a ridiculously huge PR :S

r? @oli-obk

### CODE REVIEW

File: src/tools/miri/src/shims/os_str.rs [hunk 1]
```diff
@@ -140,7 +141,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
         let size = u64::try_from(os_str.len()).unwrap().checked_add(1).unwrap(); // Make space for `0` terminator.
         let this = self.eval_context_mut();
 
-        let arg_type = this.tcx.mk_array(this.tcx.types.u8, size);
+        let arg_type = Ty::new_array(this.tcx.tcx,this.tcx.types.u8, size);
```
Comment (compiler-errors) [inline src/tools/miri/src/shims/os_str.rs:144]: subtree formatting :(

File: compiler/rustc_middle/src/ty/context.rs [hunk 2]
```diff
@@ -1845,140 +1644,29 @@ impl<'tcx> TyCtxt<'tcx> {
...
-    pub fn mk_generator_witness_mir(self, id: DefId, substs: SubstsRef<'tcx>) -> Ty<'tcx> {
-        self.mk_ty_from_kind(GeneratorWitnessMIR(id, substs))
-    }
-
     #[inline]
     pub fn mk_ct_from_kind(self, kind: ty::ConstKind<'tcx>, ty: Ty<'tcx>) -> Const<'tcx> {
         self.intern_const(ty::ConstData { kind, ty })
     }
 
+    // Avoid this in favour of more specific `mk_*` methods, where possible.
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/context.rs:1652]:
```suggestion
    // Avoid this in favour of more specific `Ty::new_*` methods, where possible.
```

**Comment by compiler-errors** [comment]
@bors r+ p=10 (very bitrotty!!)

cc subtree owners (@rust-lang/miri, @rust-lang/clippy, and @antoyo for codegen_gcc), you all will need to run rustfmt on your own after this lands.

**Comment by compiler-errors** [comment]
(oh and @bjorn3 too, you're a subtree owner too!)

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 3]
```diff
@@ -3501,7 +3502,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         {
             if let hir::Expr { kind: hir::ExprKind::Block(..), .. } = expr {
                 let expr = expr.peel_blocks();
-                let ty = typeck_results.expr_ty_adjusted_opt(expr).unwrap_or(tcx.ty_error_misc());
+                let ty = typeck_results.expr_ty_adjusted_opt(expr).unwrap_or(Ty::new_misc_error(tcx,));
```
[the remark this answers is not in the document]
Comment (BoxyUwU) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:3505]: almost certainly caused by a let else or let chain _somewhere_ in here causing rustfmt to die. Every where else formatted away the trailing `,`...

**Comment by compiler-errors** [comment]
Noting as a retrospective, but not a blocker of this PR, that we have a lot of code even within the `compiler/` directory which is unformatted and remains a bit ugly due to this PR (e.g. `Ty::mk_unit(tcx,)`). 

This is not this PR's fault, but the fault of not having formatting for let-chains in-tree. I've chatted separately with @calebcartwright to understand how we can get provisional formatting into rust sooner than later now that we (T-style) have decided on a new nightly-features formatting policy 😉

**Comment by compiler-errors** [comment]
Additional nice-to-haves after this PR lands (just the ones that come to mind):
* Replace usages of `Ty::new_unit` with `tcx.types.unit` (or vice versa) and similar with the other built-in scalar types for Consistency(TM).
* Replace usages of `ty::new_ref` that have a fixed mutability with `new_imm_ref` or `new_mut_ref` to shorten it and avoid having to go through `ty::TyAndMutability` -- `new_ref` should imo only be used when the mutability is being passed in or is variable or something.
* Ideally we'd fix some of the annoying formatting blips (like a lack of spacing after `tcx,` in the args) before getting full let-chains support...

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (calebcartwright) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:3505]: yeah there's a containing let chain a couple lines up. obviously wouldn't have helped with this particular line, but fwiw, now that let-else formatting is available, the pinned version of rustfmt used in this repo could be updated to a newer rustfmt version to eliminate let-else as a source of skipped formatting.

unsure how a big a "format the world" churn, let-else edition, that'll entail though :smile:

**Pull request merged**